Line/android/LINELoginDemo/app/src/main/java/com/google/firebase/linelogindemo/util/LineLoginHelper.java

```diff
@@ -0,0 +1,255 @@
+        auth.signInWithCustomToken(firebaseToken)
+                .addOnCompleteListener(activity, new OnCompleteListener<AuthResult>() {
+                    @Override
+                    public void onComplete(@NonNull Task<AuthResult> task) {
+                        Log.d(TAG, "signInWithCustomToken:onComplete:" + task.isSuccessful());
+                        if (task.isSuccessful()) {
+                            source.setResult(task.getResult().getUser());
+                        } else {
+                            source.setException(task.getException());
+                        }
+                    }
+                });
+
+        return source.getTask();
+    }
+
+    private Task<FirebaseUser> updateFirebaseUserProfile(Activity activity, final FirebaseUser user) {
+        final TaskCompletionSource<FirebaseUser> source = new TaskCompletionSource<>();
+
+        // Check if we should use LINE profile to update Firebase profile
+        boolean isProfileNeedUpdate = (user.getDisplayName() == null) || (user.getPhotoUrl() == null);
+        if (!isProfileNeedUpdate) {
+            // Return the current user immediately
+            source.setResult(user);
+            return source.getTask();
+        }
+
+        // STEP 4: (Optional) Update user profile with LINE Profile
+        ApiClient apiClient = LineSdkContextManager.getSdkContext().getApiClient();
+        apiClient.getMyProfile(new ApiRequestFutureListener<Profile>() {
+            @Override
+            public void requestComplete(ApiRequestFuture<Profile> future) {
+                if (future.getStatus() != FutureStatus.SUCCESS) {
+                    // As updating user profile is an optional step, return the user regardless of the result
+                    source.setResult(user);
+                    return;
+                }
+
+                // Get LINE Profile
+                Profile profile = future.getResponseObject();
+                String displayName = profile.displayName;
+                String pictureUrl = profile.pictureUrl;
+
+                // Update Firebase profile
+                UserProfileChangeRequest changeRequest =
+                        new UserProfileChangeRequest.Builder()
+                                .setDisplayName(displayName)
+                                .setPhotoUri(Uri.parse(pictureUrl))
+                                .build();
+                user.updateProfile(changeRequest).addOnCompleteListener(new OnCompleteListener<Void>() {
+                    @Override
+                    public void onComplete(@NonNull Task<Void> task) {
+                        // As updating user profile is an optional step, return the user regardless of the result
+                        source.setResult(user);
+                    }
+                });
+            }
+        });
+
+        return source.getTask();
+    }
+
+    public void signOut() {
+        FirebaseAuth.getInstance().signOut();
```

Comment:
Ignore my previous comment, but I'd just inline this method in the Activity so it's clearer what's being done without jumping between files.  Seems like this method does not need anything from this class.